cpp/src/copying/contiguous_split.cu

```diff
@@ -0,0 +1,219 @@
+#include <cudf/copying.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+
+namespace cudf {
+
+namespace experimental {
+
+namespace detail {
+
+namespace {
+
+template <size_type block_size, typename T, bool has_validity>
+__launch_bounds__(block_size)
+__global__
+void copy_in_place_kernel( column_device_view const in,
+                           mutable_column_device_view out)
+{
+   const size_type tid = threadIdx.x + blockIdx.x * block_size;
+   const int warp_id = tid / cudf::experimental::detail::warp_size;
+   const size_type warps_per_grid = gridDim.x * block_size / cudf::experimental::detail::warp_size;      
+
+   // begin/end indices for the column data
+   size_type begin = 0;
+   size_type end = in.size();
+   // warp indices.  since 1 warp == 32 threads == sizeof(bit_mask_t) * 8,
+   // each warp will process one (32 bit) of the validity mask via
+   // __ballot_sync()
+   size_type warp_begin = cudf::word_index(begin);
+   size_type warp_end = cudf::word_index(end-1);      
+
+   // lane id within the current warp   
+   const int lane_id = threadIdx.x % cudf::experimental::detail::warp_size;
+   
+   // current warp.
+   size_type warp_cur = warp_begin + warp_id;   
+   size_type index = tid;
+   while(warp_cur <= warp_end){
+      bool in_range = (index >= begin && index < end);
+            
+      bool valid = true;
+      if(has_validity){
+         valid = in_range && in.is_valid(index);
+      }
+      if(in_range){
+         out.element<T>(index) = in.element<T>(index);
+      }
+      
+      // update validity      
+      if(has_validity){
+         // the final validity mask for this warp 
+         int warp_mask = __ballot_sync(0xFFFF'FFFF, valid && in_range);
+         // only one guy in the warp needs to update the mask and count
+         if(lane_id == 0){            
+            out.set_mask_word(warp_cur, warp_mask);            
+         }
+      }            
+
+      // next grid
+      warp_cur += warps_per_grid;
+      index += block_size * gridDim.x;
+   }      
+}
+
+
+static constexpr size_t split_align = 8;
+
+template<typename T>
+struct column_buf_size_functor_impl {
+   void operator()(column_view const& c, size_t& running_data_size, size_t& running_validity_size)
+   {
+      running_data_size += cudf::util::div_rounding_up_safe(c.size() * sizeof(T), split_align) * split_align;      
+      if(c.nullable()){
+         running_validity_size += cudf::bitmask_allocation_size_bytes(c.size(), split_align);         
+      }
+   }
+};
+
+template<>
+struct column_buf_size_functor_impl<string_view> {
+   void operator()(column_view const& c, size_t& running_data_size, size_t& running_validity_size)
+   {
+      CUDF_FAIL("contiguous_split for strings not implemented yet");
+   };
+};
+
+struct column_buf_size_functor {
+   template<typename T>
+   void operator()(column_view const& c, size_t& running_data_size, size_t& running_validity_size)
+   {
+      column_buf_size_functor_impl<T> sizer{};
```

Comment:
Same comment here. `column_buf_size_functor::operator()` can be specialized, which obviates the need for the extra indirection of `column_buf_size_functor_impl`.